### PR TITLE
Deyahpify cifar10

### PR DESCRIPTION
**Pull request by growlix**
# What does this PR do?

This PR completes the deyahpification of CIFAR10 by implementing a factory method `build_ffcv_cifar10_dataloader()`.

# What issue(s) does this change relate to?

https://mosaicml.atlassian.net/browse/RESEARCH-290
https://mosaicml.atlassian.net/jira/software/c/projects/CO/boards/12/roadmap?selectedIssue=CO-1000

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by dskhudia — APPROVED** [review]

File: composer/datasets/cifar.py [hunk 1]
```diff
@@ -80,6 +86,106 @@ def build_cifar10_dataloader(
...
+        download (bool, optional): Whether to download the dataset, if needed. Default:
+            ``True``.
+        drop_last (bool): Drop remainder samples. Default: ``True``.
+        drop_last (bool): whether to drop last samples. Default: ``True``.
+        prefetch_factor (int): Number of batches to prefect. Default: ``2``.
+        ffcv_dir (str, optional): A directory containing train/val <file>.ffcv files. If these files don't exist and
+            ``ffcv_write_dataset`` is ``True``, train/val <file>.ffcv files will be created in this dir. Default: ``"/tmp"``.
+        ffcv_dest (str, optional): <file>.ffcv file that has dataset samples. Default: ``"cifar_train.ffcv"``.
+        ffcv_write_dataset (str | bool, optional): Whether to create dataset in FFCV format (<file>.ffcv) if it doesn't exist. Default:
+        ``False``.
```
Comment (knighton) [inline composer/datasets/cifar.py:116]: missing indent (screws up docs)

File: composer/datasets/cifar.py [hunk 2]
```diff
@@ -80,6 +86,106 @@ def build_cifar10_dataloader(
...
+):
+    """Builds and FFCV CIFAR10 dataloader
+
+    Args:
+        batch_size (int): Batch size per device
+        is_train (bool): Whether to load the training data or validation data. Default:
+            ``True``.
+        download (bool, optional): Whether to download the dataset, if needed. Default:
+            ``True``.
+        drop_last (bool): Drop remainder samples. Default: ``True``.
```
Comment (knighton) [inline composer/datasets/cifar.py:109]: drop_last is double listed

File: composer/datasets/cifar.py [hunk 3]
```diff
@@ -80,6 +86,106 @@ def build_cifar10_dataloader(
...
+    num_workers: int = 8,
+    ffcv_dir: str = '/tmp',
+    ffcv_dest: str = 'cifar_train.ffcv',
+    ffcv_write_dataset: Union[str, bool] = False,
+    datadir: Union[str, None] = None,
+):
+    """Builds and FFCV CIFAR10 dataloader
+
+    Args:
+        batch_size (int): Batch size per device
```
Comment (knighton) [inline composer/datasets/cifar.py:104]: missing trailing .

File: composer/datasets/cifar.py [hunk 4]
```diff
@@ -80,6 +86,106 @@ def build_cifar10_dataloader(
...
+            ``True``.
+        drop_last (bool): Drop remainder samples. Default: ``True``.
+        drop_last (bool): whether to drop last samples. Default: ``True``.
+        prefetch_factor (int): Number of batches to prefect. Default: ``2``.
+        ffcv_dir (str, optional): A directory containing train/val <file>.ffcv files. If these files don't exist and
+            ``ffcv_write_dataset`` is ``True``, train/val <file>.ffcv files will be created in this dir. Default: ``"/tmp"``.
+        ffcv_dest (str, optional): <file>.ffcv file that has dataset samples. Default: ``"cifar_train.ffcv"``.
+        ffcv_write_dataset (str | bool, optional): Whether to create dataset in FFCV format (<file>.ffcv) if it doesn't exist. Default:
+        ``False``.
+        datadir (str | None, optional): Path to the non-FFCV data directory
```
Comment (knighton) [inline composer/datasets/cifar.py:117]: missing trailing .

File: composer/datasets/cifar.py [hunk 5]
```diff
@@ -80,6 +86,106 @@ def build_cifar10_dataloader(
...
+        dist.barrier()
+
+    if not os.path.exists(dataset_filepath):
+        raise ValueError(
+            f'Dataset file containing samples not found at {dataset_filepath}. Use ffcv_dir flag to point to a dir containing {dataset_filepath}.'
+        )
+
+    # Please note that this mean/std is different from the mean/std used for regular PyTorch dataloader as
+    # ToTensor does the normalization for PyTorch dataloaders.
+    cifar10_mean_ffcv = [125.307, 122.961, 113.8575]
```
Comment (knighton) [inline composer/datasets/cifar.py:153]: superfluous brackets (?)

File: composer/datasets/cifar.py [hunk 6]
```diff
@@ -80,6 +86,106 @@ def build_cifar10_dataloader(
...
+
+    return ffcv.Loader(
+        dataset_filepath,
+        batch_size=batch_size,
+        num_workers=num_workers,
+        order=ordering,
+        distributed=False,
+        pipelines={
+            'image': image_pipeline,
+            'label': label_pipeline
```
Comment (knighton) [inline composer/datasets/cifar.py:182]: no trailing comma here, but does have trailing comma on :185

**Review by knighton — APPROVED** [review]
who you calling a dummy?

File: composer/datasets/cifar.py [hunk 7]
```diff
@@ -80,6 +86,106 @@ def build_cifar10_dataloader(
...
+    download: bool = True,
+    drop_last: bool = True,
+    prefetch_factor: int = 2,
+    num_workers: int = 8,
+    ffcv_dir: str = '/tmp',
+    ffcv_dest: str = 'cifar_train.ffcv',
+    ffcv_write_dataset: Union[str, bool] = False,
+    datadir: Union[str, None] = None,
+):
+    """Builds and FFCV CIFAR10 dataloader
```
Comment (knighton) [inline composer/datasets/cifar.py:101]: s/and/an/

missing trialing .

File: composer/datasets/cifar.py (hunk 5 above)
Comment (growlix) [reply composer/datasets/cifar.py:153]: I'm going to assume it's supposed to be a list and not a tuple and keep those

**Pull request merged**